### PR TITLE
REF: stronger typing in _box_func

### PR DESCRIPTION
**Pull request by jbrockmendel**
Necessary for non-nano support

### CODE REVIEW

File: pandas/core/nanops.py [hunk 1]
```diff
@@ -367,19 +366,23 @@ def _wrap_results(result, dtype: np.dtype, fill_value=None):
                 result = np.datetime64("NaT", "ns")
             else:
                 result = np.int64(result).view("datetime64[ns]")
+            # retain original unit
+            result = result.astype(dtype, copy=False)
```
Comment (mroeschke) [inline pandas/core/nanops.py:370]: So do these changes make things internally More Correct and not necessarily have an external behavior change?

Comment (jbrockmendel) [reply pandas/core/nanops.py:370]: not user-facing unless they are directly using nanops, which they shouldnt be

File: pandas/core/arrays/datetimes.py [hunk 2]
```diff
@@ -538,13 +538,10 @@ def _check_compatible_with(self, other, setitem: bool = False):
...
     # Descriptive Properties
 
-    def _box_func(self, x) -> Timestamp | NaTType:
-        if isinstance(x, np.datetime64):
-            # GH#42228
-            # Argument 1 to "signedinteger" has incompatible type "datetime64";
-            # expected "Union[SupportsInt, Union[str, bytes], SupportsIndex]"
-            x = np.int64(x)  # type: ignore[arg-type]
-        ts = Timestamp(x, tz=self.tz)
+    def _box_func(self, x: np.datetime64) -> Timestamp | NaTType:
```
Comment (jreback) [inline pandas/core/arrays/datetimes.py:541]: is this guaranteed true? e.g. can the input be a Timestamp

Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py:541]: the nanops changes fix the only places where we call _box_func with incorrect scalars

**Pull request merged**